File: Problem572/Solution.java
/**
 * @author tgorthi
 * @since Jun 2020
 */
public class Solution
{
    public boolean isSubtree(TreeNode s, TreeNode t)
    {
        return isSubtree(s, t, false);
    }

    private boolean isSubtree(TreeNode s, TreeNode t, boolean partialMatch)
    {
        if ((s == null && t != null) || (s != null && t == null))
        {
            return false;
        }
        if (s == null && t == null)
        {
            return true;
        }
        if (t.val == s.val && isSubtree(s.left, t.left, true) && isSubtree(s.right, t.right, true))
        {
            return true;
        }
        if (partialMatch)
        {
            return false;
        }
        return isSubtree(s.left, t, false) || isSubtree(s.right, t, false);
    }
}

File: Problem572/Solution.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSubtree(s *TreeNode, t *TreeNode) bool {
	return checkIfSubtree(s, t, false)
}

func checkIfSubtree(s *TreeNode, t *TreeNode, isPartial bool) bool {
	if (s == nil && t != nil) || (s != nil && t == nil) {
		return false
	}
	if s == nil && t == nil {
		return true
	}
	if t.Val == s.Val && checkIfSubtree(s.Left, t.Left, true) && checkIfSubtree(s.Right, t.Right, true) {
		return true
	}

	if isPartial {
		return false
	}
	return checkIfSubtree(s.Left, t, false) || checkIfSubtree(s.Right, t, false)
}
